event/datadeal/data_verification_event.go

```diff
@@ -1 +1,216 @@
 package datadeal
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"strconv"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/medibloc/panacea-core/v2/x/datadeal/types"
+	oracletypes "github.com/medibloc/panacea-core/v2/x/oracle/types"
+	"github.com/medibloc/panacea-doracle/config"
+	"github.com/medibloc/panacea-doracle/crypto"
+	"github.com/medibloc/panacea-doracle/event"
+	"github.com/medibloc/panacea-doracle/ipfs"
+	"github.com/medibloc/panacea-doracle/panacea"
+	"github.com/medibloc/panacea-doracle/sgx"
+	"github.com/medibloc/panacea-doracle/validation"
+	log "github.com/sirupsen/logrus"
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+)
+
+var _ event.Event = (*DataVerificationEvent)(nil)
+
+type DataVerificationEvent struct {
+	reactor reactor
+}
+
+type reactor interface {
+	GRPCClient() *panacea.GrpcClient
+	EnclaveInfo() *sgx.EnclaveInfo
+	OracleAcc() *panacea.OracleAccount
+	OraclePrivKey() *btcec.PrivateKey
+	Config() *config.Config
+	QueryClient() *panacea.QueryClient
+	Ipfs() *ipfs.Ipfs
+}
+
+func NewDataVerificationEvent(r reactor) DataVerificationEvent {
+	return DataVerificationEvent{r}
+}
+
+func (d DataVerificationEvent) GetEventType() string {
+	return "message"
+}
+
+func (d DataVerificationEvent) GetEventAttributeKey() string {
+	return "action"
+}
+
+func (d DataVerificationEvent) GetEventAttributeValue() string {
+	return "'SellData'"
+}
+
+func (d DataVerificationEvent) EventHandler(event ctypes.ResultEvent) error {
+	dealIDStr := event.Events[types.EventTypeDataVerificationVote+"."+types.AttributeKeyDealID][0]
+	dataHash := event.Events[types.EventTypeDataVerificationVote+"."+types.AttributeKeyDataHash][0]
+
+	dealID, err := strconv.ParseUint(dealIDStr, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	deal, err := d.reactor.QueryClient().GetDeal(dealID)
+	if err != nil {
+		return err
+	}
+
+	dataSale, err := d.reactor.QueryClient().GetDataSale(dataHash, dealID)
+	if err != nil {
+		return err
+	}
+
+	encryptedDataBz, err := d.reactor.Ipfs().Get(dataSale.VerifiableCid)
+	if err != nil {
+		return err
+	}
+
+	oraclePrivKey := d.reactor.OraclePrivKey()
+
+	sellerAcc, err := d.reactor.QueryClient().GetAccount(dataSale.SellerAddress)
+	if err != nil {
+		return err
+	}
+
+	sellerPubKeyBytes := sellerAcc.GetPubKey().Bytes()
+
+	sellerPubKey, err := btcec.ParsePubKey(sellerPubKeyBytes, btcec.S256())
+	if err != nil {
+		return err
+	}
+
+	decryptSharedKey := crypto.DeriveSharedKey(oraclePrivKey, sellerPubKey, crypto.KDFSHA256)
+
+	decryptedData, err := d.decryptData(decryptSharedKey, deal.Nonce, encryptedDataBz)
+	if err != nil {
+		return err
+	}
```

Comment:
If decryption fails, doracle may vote `NO` as well. I think it's invalid data sale.